presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java

```diff
@@ -0,0 +1,224 @@
+                        throw e;
+                    }
+                }
+            }
+        }
+        throw new PrestoException(FUNCTION_NOT_FOUND, format("Function %s not found in path: %s", name, session.getPath()));
+    }
+
+    public WindowFunctionSupplier getWindowFunctionImplementation(Signature signature)
+    {
+        try {
+            return operatorNamespace.getWindowFunctionImplementation(signature);
+        }
+        catch (PrestoException e) {
+            //do nothing and try to resolve normally
+        }
+        if (!signature.getCatalog().isPresent() || !signature.getSchema().isPresent()) {
+            throw new PrestoException(NOT_FOUND, format("Signature %s does not have both catalog and schema set for resolution", signature));
+        }
+        return functionNamespaces.get(signature.getCatalog().get()).getWindowFunctionImplementation(signature);
+    }
+
+    public InternalAggregationFunction getAggregateFunctionImplementation(Signature signature)
+    {
+        try {
+            return operatorNamespace.getAggregateFunctionImplementation(signature);
+        }
+        catch (PrestoException e) {
+            //do nothing and try to resolve normally
+        }
+        if (!signature.getCatalog().isPresent() || !signature.getSchema().isPresent()) {
+            throw new PrestoException(NOT_FOUND, format("Signature %s does not have both catalog and schema set for resolution", signature));
+        }
+        return functionNamespaces.get(signature.getCatalog().get()).getAggregateFunctionImplementation(signature);
+    }
+
+    public ScalarFunctionImplementation getScalarFunctionImplementation(Signature signature)
+    {
+        try {
+            return operatorNamespace.getScalarFunctionImplementation(signature);
+        }
+        catch (PrestoException e) {
+            //do nothing and try to resolve normally
+        }
+        if (!signature.getCatalog().isPresent() || !signature.getSchema().isPresent()) {
+            throw new PrestoException(NOT_FOUND, format("Signature %s does not have both catalog and schema set for resolution", signature));
+        }
+        return functionNamespaces.get(signature.getCatalog().get()).getScalarFunctionImplementation(signature);
+    }
+
+    public boolean canResolveOperator(OperatorType operatorType, Type returnType, List<? extends Type> argumentTypes)
+    {
+        return operatorNamespace.canResolveOperator(operatorType, returnType, argumentTypes);
+    }
+
+    public Signature resolveOperator(OperatorType operatorType, List<? extends Type> argumentTypes)
+            throws OperatorNotFoundException
+    {
+        try {
+            return operatorNamespace.resolveOperator(operatorType, argumentTypes);
+        }
+        catch (PrestoException e) {
+            if (!(e instanceof OperatorNotFoundException)) {
+                throw e;
+            }
+        }
+        throw new OperatorNotFoundException(
```

Comment:
Why don't we throw `e` above when it is an instance of `OperatorNotFoundException`? Doesn't it already include the info we need?